### PR TITLE
[Do not merge][graphbolt] s3-fifo-cache for DiskBasedFeature

### PR DESCRIPTION
**Pull request by pyynb**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by mfbalin** [comment]
Is the data required to run this code generated by some other code?

**Comment by pyynb** [comment]
> Is the data required to run this code generated by some other code?

Yes, during a training process, I recorded all the indexes and saved them in a txt file.
The file is too large so i can't push it. You can try recording it yourself.

**Comment by mfbalin** [comment]
@pyynb Could you run the cache hit rate benchmark with the cache in #7492? It can be accessed via `gb.impl.FeatureCache`.

File: cache.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        if key in self.cache:
+            self.hit += 1
+            self.cache[key] = value
+            self.cache.move_to_end(key)
+        else:
+            if len(self.cache) >= self.capacity:
+                self.cache.popitem(last=False)
+            self.cache[key] = value
+
+
```
Comment (mfbalin) [inline cache.py:31]:
```suggestion
class GraphBoltS3FIFO:
    def __init__(self, number_of_elements: int, num_parts: int):
        self.cache = gb.impl.FeatureCache([number_of_elements, 1], torch.int64, num_parts)

    def get(self, key: int) -> int:
        values, missing_index, missing_keys = self.cache.query(torch.tensor([key]))
        if missing_keys.numel() == 0:
            return values[0].item()
        return -1

    def put(self, key: int, value: int) -> None:
        self.cache.replace(torch.tensor([key]), torch.tensor([value]))

    @property
    def hit_rate(self) -> float:
        return 1 - self.cache.miss_rate
```

Comment (mfbalin) [reply cache.py:31]: num_parts is there to add parallelism. It would be nice if we could measure num_parts == 1 and num_parts == 16.

Comment (pyynb) [reply cache.py:31]: i see. thank you for your advice!

**Comment by mfbalin** [comment]
https://github.com/dmlc/dgl/tree/master/examples/graphbolt/disk_based_feature

We have experimental results comparing different caching policies here.

**Pull request closed without merging**